File: maximumErasureValue.py
# problem: https: // leetcode.com/problems/maximum-erasure-value/
# Runtime: 2756 ms, faster than 6.09% of Python3 online submissions for Maximum Erasure Value.
# Memory Usage: 28.1 MB, less than 63.08% of Python3 online submissions for Maximum Erasure Value.
from typing import List


class Solution:
    def maximumUniqueSubarray(self, nums: List[int]) -> int:
        from collections import deque, defaultdict
        dp = {}
        start = 0
        end = 0
        answer = 0
        for i in range(len(nums)):
            if nums[i] in dp:
                for j in range(start, dp[nums[i]]):
                    del dp[nums[j]]
                start = dp[nums[i]] + 1
            else:
                end = i
            dp[nums[i]] = i
            answer = max(answer, sum(nums[start:end + 1]))
        return answer

File: maximumErasureValue.go
package algorithms

// problem: https://leetcode.com/problems/maximum-erasure-value/submissions/
// Runtime: 189 ms, faster than 92.00% of Go online submissions for Maximum Erasure Value.
// Memory Usage: 10.2 MB, less than 20.00% of Go online submissions for Maximum Erasure Value.

func maximumUniqueSubarray(nums []int) int {
	seen := make([]bool, 10e5+1)
	i, j := 0, 0
	n := len(nums)
	answer := 0
	cur := 0
	for i < n && j < n {
		if !seen[nums[j]] {
			cur += nums[j]
			seen[nums[j]] = true
			j++
			if cur > answer {
				answer = cur
			}
		} else {
			cur -= nums[i]
			seen[nums[i]] = false
			i++
		}
	}
	return answer
}
